sycl/source/detail/event_impl.cpp

```diff
@@ -174,21 +175,31 @@ event_impl::event_impl(ur_event_handle_t Event, const context &SyclContext)
   }
 }
 
-event_impl::event_impl(const QueueImplPtr &Queue)
-    : MQueue{Queue}, MIsProfilingEnabled{!Queue || Queue->MIsProfilingEnabled} {
-  if (Queue)
-    this->setContextImpl(Queue->getContextImplPtr());
-  else {
-    MState.store(HES_NotComplete);
+event_impl::event_impl(queue_impl &Queue, private_tag)
+    : MQueue{Queue.weak_from_this()},
+      MIsProfilingEnabled{Queue.MIsProfilingEnabled} {
+  this->setContextImpl(Queue.getContextImplPtr());
+  MState.store(HES_Complete);
+}
+
+event_impl::event_impl(HostEventState State, private_tag) : MState(State) {
+  switch (State) {
+  case HES_Discarded:
+  case HES_Complete: {
+    MIsFlushed = true;
+    MIsHostEvent = true;
+    break;
+  }
+  case HES_NotComplete: {
+    MIsProfilingEnabled = true;
```

Comment:
It looks like `MIsHostEvent` wasn't set for this scenario before, but this is weird to me. If anyone has any idea, I'd like to hear it.

Reply:
Sadly I don't remember the details either, but it smells like events are due for some refactoring in general. The idea of "host" seems very overloaded.